pyro/contrib/gp/models/gpr.py

```diff
@@ -5,81 +5,94 @@
+        pyro.sample("y", dist.MultivariateNormal(zero_loc, K), obs=self.y)
 
     def guide(self):
-        guide_priors = {}
-        for p in self.priors:
+        kernel_guide_prior = {}
+        for p in self.kernel_prior:
             p_MAP_name = pyro.param_with_module_name(self.kernel.name, p) + "_MAP"
             # init params by their prior means
-            p_MAP = pyro.param(p_MAP_name, Variable(self.priors[p].analytic_mean().data.clone(),
+            p_MAP = pyro.param(p_MAP_name, Variable(self.kernel_prior[p].torch_dist.mean.data.clone(),
                                                     requires_grad=True))
-            guide_priors[p] = dist.Delta(p_MAP)
-        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, guide_priors)
-        return kernel_fn()
+            kernel_guide_prior[p] = dist.Delta(p_MAP)
 
-    def forward(self, Z):
+        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, kernel_guide_prior)
+        kernel = kernel_fn()
+
+        return kernel
+
+    def forward(self, Xnew, full_cov=False, noiseless=False):
         """
-        Compute the parameters of `p(y|Z) ~ N(loc, covariance_matrix)`
-            w.r.t. the new input Z.
+        Compute the parameters of `p(y|Xnew) ~ N(loc, cov)`
+            w.r.t. the new input Xnew.
 
-        :param torch.autograd.Variable Z: A 2D tensor.
-        :return: loc and covariance matrix of p(y|Z).
+        :param torch.autograd.Variable Xnew: A 2D tensor.
+        :return: loc and covariance matrix of p(y|Xnew).
         :rtype: torch.autograd.Variable and torch.autograd.Variable
         """
-        if Z.dim() == 2 and self.X.size(1) != Z.size(1):
+        if Xnew.dim() == 2 and self.X.size(1) != Xnew.size(1):
             assert ValueError("Train data and test data should have the same feature sizes.")
-        if Z.dim() == 1:
-            Z = Z.unsqueeze(1)
+        if Xnew.dim() == 1:
+            Xnew = Xnew.unsqueeze(1)
+
         kernel = self.guide()
-        K = kernel(self.X) + self.noise.repeat(self.input_dim).diag()
+        Kff = kernel(self.X)
+        if noiseless:
+            Kff = Kff + self.jitter.expand(self.num_data).diag()
+        else:
+            Kff = Kff + self.noise.expand(self.num_data).diag()
+        Kfs = kernel(self.X, Xnew)
+        Lff = Kff.potrf(upper=False)
 
-        K_xz = kernel(self.X, Z)
-        K_zx = K_xz.t()
-        K_zz = kernel(Z)
+        Lffinv_y = matrix_triangular_solve_compat(self.y, Lff, upper=False)
+        Lffinv_Kfs = matrix_triangular_solve_compat(Kfs, Lff, upper=False)
 
-        # TODO Use torch.trtrs or torch.potrs when it supports cuda tensors
-        # and is differentiable.
-        # Refer to Gaussian processes for machine learning main gpr algorithm
-        # L = torch.potrf(K, upper=False)
-        # alpha = torch.potrs(L, self.y, upper=False)
+        # loc = Kfs.T @ inv(Kff) @ y
+        loc = Lffinv_Kfs.t().matmul(Lffinv_y)
```

Comment:
wow, cool, this works now?

Reply:
For `torch.trtrs` cuda and grad: not yet. We make a wrapper here: https://github.com/uber/pyro/blob/dev/pyro/distributions/util.py#L274